File: PA118.java
//Leetcode : 189. Rotate Array
//Time Complexity: O(n) , n is the size of the array
//Space Complexity; O(1), No extra space used
class Solution {
    public void rotate(int[] nums, int k) {
        if(nums==null || nums.length==0) return;
        k=k%nums.length;
         reverse(nums,0,nums.length-1);       
         reverse(nums,k,nums.length-1);
        reverse(nums,0,k-1);
        
    }
    
    public int[] reverse(int[] arr,int i, int j){
        while(i<j){
            int temp =arr[i];
            arr[i]=arr[j];
            arr[j]= temp;
            i++;
            j--;
        }
        return arr;
    }
}

File: PA118.py
#Time Complexity - O(n)
#Space Complexity - O(n)
class Solution(object):
    def rotate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        tmp = [0] * len(nums)
        print(tmp)
        for i in range(len(nums)):
            #print(tmp[(i+k) % len(nums)])
            tmp[(i+k)%len(nums)] = nums[i]
        
        for i in range(len(nums)):
            nums[i] = tmp[i]